application/src/main/java/org/opentripplanner/graph_builder/module/osm/EdgeLevelInfoProcessor.java

```diff
@@ -0,0 +1,120 @@
+package org.opentripplanner.graph_builder.module.osm;
+
+import java.util.List;
+import java.util.Optional;
+import javax.annotation.Nullable;
+import org.opentripplanner.graph_builder.issue.api.DataImportIssueStore;
+import org.opentripplanner.graph_builder.issues.CouldNotApplyMultiLevelInfoToWay;
+import org.opentripplanner.osm.model.OsmLevel;
+import org.opentripplanner.osm.model.OsmWay;
+import org.opentripplanner.service.streetdecorator.OsmStreetDecoratorRepository;
+import org.opentripplanner.service.streetdecorator.model.EdgeLevelInfo;
+import org.opentripplanner.service.streetdecorator.model.Level;
+import org.opentripplanner.service.streetdecorator.model.VertexLevelInfo;
+import org.opentripplanner.street.model.edge.Edge;
+
+/**
+ * Contains logic for storing edge level info in the
+ * {@link OsmStreetDecoratorRepository}.
+ */
+class EdgeLevelInfoProcessor {
+
+  private final DataImportIssueStore issueStore;
+  private final OsmStreetDecoratorRepository osmStreetDecoratorRepository;
+  private final boolean includeEdgeLevelInfo;
+
+  public EdgeLevelInfoProcessor(
+    DataImportIssueStore issueStore,
+    boolean includeEdgeLevelInfo,
+    OsmStreetDecoratorRepository osmStreetDecoratorRepository
+  ) {
+    this.issueStore = issueStore;
+    this.includeEdgeLevelInfo = includeEdgeLevelInfo;
+    this.osmStreetDecoratorRepository = osmStreetDecoratorRepository;
+  }
+
+  public Optional<EdgeLevelInfo> getEdgeLevelInfo(OsmDatabase osmdb, OsmWay way) {
+    if (!includeEdgeLevelInfo) {
+      return Optional.empty();
+    }
+
+    List<OsmLevel> levels = osmdb.getLevelsForEntity(way);
+    var nodeRefs = way.getNodeRefs();
+    if (nodeRefs.size() > 1) {
```

Comment:
You could check that node nodeRefs.size() < 2 as a precondition at the beginning of the method so this would be less nested.

Reply:
I removed the check entirely, I think in other places in the codebase there is an assumption that ways have 2 or more nodes. I'm not sure if this is checked anywhere though